# C 错误处理

C 语言不提供对错误处理的直接支持，但是作为一种系统编程语言，它以返回值的形式允许您访问底层数据。
在发生错误时，大多数的 C 或 UNIX 函数调用返回 1 或 NULL，同时会设置一个错误代码 errno，
该错误代码是全局变量，表示在函数调用期间发生了错误

## `errno`, `perror()`, `strerror()`

- `perror()` : 函数显示您传给它的字符串，后跟一个冒号、一个空格和当前 `errno` 值的文本表示形式。
- `strerror()` : 返回一个指针, 指针指向当前 `errno` 值的文本表示形式

```
#include <stdio.h>
#include <errno.h>
#include <string.h>

extern int errno;


int main(int argc, char const *argv[])
{
    FILE *fp;
    int errnum;
    fp = fopen("a.txt", "rb");

    if(fp == NULL){

        errnum = errno;
        fprintf(stderr, "错误号: %d\n", errno);
        perror("通过 perror 输出错误");
        fprintf(stderr, "打开文件错误: %s\n", strerror(errnum));

    } else {
        fclose(fp);
    }

    return 0;
}

```

## 程序退出状态

宏定义:
- EXIT_SUCCESS : 0
- EXIT_FAILURE : -1

```
#include <stdio.h>
#include <stdlib.h>


int main(int argc, char const *argv[])
{
    int dividend = 20;
    int divisor = 0;
    int quotient;

    if( divisor == 0){
        fprintf(stderr, "除数为 0 退出运行...\n");
        exit(EXIT_FAILURE);
    }

    quotient = dividend / divisor;
    fprintf(stderr, "quotient 变量的值为 : %d\n", quotient );

    exit(EXIT_SUCCESS);
    return 0;
}

```
